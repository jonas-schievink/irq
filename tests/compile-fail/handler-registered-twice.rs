//! Tests that a handler can only be registered once for the duration of a `scope`.
//!
//! This is required for safety since otherwise the handler could preempt itself and duplicate
//! mutable references during the second invocation.

use irq::{handler, scope, scoped_interrupts};
use mock_pac::interrupt;

scoped_interrupts! {
    enum Interrupt {
        INT0,
        INT1,
    }

    use #[interrupt];
}

fn main() {
    let mut i = 0;
    handler!(int0 = move || i += 1);

    scope(|scope| {
        scope.register(Interrupt::INT0, int0);
        scope.register(Interrupt::INT1, int0);

        loop {
            // Idle loop
            break;
        }
    });
}
