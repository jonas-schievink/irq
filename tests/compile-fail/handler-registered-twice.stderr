error[E0499]: cannot borrow `*int0` as mutable more than once at a time
  --> $DIR/handler-registered-twice.rs:24:41
   |
23 |         scope.register(Interrupt::INT0, int0);
   |                                         ---- first mutable borrow occurs here
24 |         scope.register(Interrupt::INT1, int0);
   |               --------                  ^^^^ second mutable borrow occurs here
   |               |
   |               first borrow later used by call
